New portion of linux fixes.

eagleivg opened the pull request:
New portion of linux fixes.  Major cross-platform changes:

- fixed console command vid_mode, vid_refresh and e.t.c. (overloading virtual methods - pure evil)

- Add code for splash window via SDL2 (linux only yet)
- Fix broken priority in messages queue

q4a committed:
add -fvisibility=hidden to CMake

Zegeri committed:
Reduce the amount of warnings

* Inlined pure virtual functions don't make sense.
* Use '= delete' to prohibit operators instead of runtime errors.
* Fix mistakes in return values.

q4a committed:
Fix fvisibility

Zegeri committed:
xrPhysics: Export symbol ph_world

q4a committed:
Compare size with 0, not NULL

drug007 committed:
Diagnostic improved.

Zegeri committed:
Fix _splitpath returning incorrect extension

_splitpath is supposed to return the dot before the extension.

Zegeri committed:
xrEngine: Fix integer overflow of Pure classes

Also use std::sort, which is faster than qsort.

Zegeri committed:
Manage gStringTable creation/destruction in Linux

eagleivg committed:
common: improve linux snprintf_s implementation (fix wrong processing console text in line_edit_control::add_inserted_text)

eagleivg committed:
xrScriptEngine: add debug defines under comment

eagleivg committed:
xrGame: fix dependency of CUIDialogWndEx under Linux

eagleivg committed:
xrGameSpy, GameSpy: added to linux build

eagleivg committed:
xrRenderGL: fix gl texture loading on linux

eagleivg committed:
xrGame: option -nointro now skip intro video

eagleivg committed:
xrRender: fix model loading on linux

drug007 committed:
add asserts

eagleivg committed:
xrCore, xrRender: more linux delimiter replacement

Zegeri committed:
xrGame: Resolve mixed_delegate::bind's ambiguous type for luabind

eagleivg committed:
GameSpy: fixed undefined reference ptCheckForPatchA

eagleivg committed:
xrGame: fix undefined reference of agent_manager()

eagleivg committed:
xrGame: fix undefined reference to `CScriptGameObject::object() const'

eagleivg committed:
xrServerEntities: fix undefined reference to `CALifeMonsterBrain::smart_terrain()

eagleivg committed:
xrGame: fix undefined reference to gamespy_profile::stats_submitter...

eagleivg committed:
xrEngine: fix undefined reference to `CObjectList::o_crow(IGameObject*)'

q4a committed:
Fix warning: "PURE_DYNAMIC_CAST" redefined

eagleivg committed:
xrCore: fixed linux Event implementation according Windows

eagleivg committed:
xrGame: implement linux GetCurrentProcessId()

eagleivg committed:
xrGame: implement linux _InterlockedCompareExchange()

eagleivg committed:
xrGame: implement linux strncat_s()

eagleivg committed:
xrGame: ename implemented SwitchToThread() calls

eagleivg committed:
xrGame: fixed 2 more undefined reference on linux

eagleivg committed:
linux: refactor some delimiter repaces

eagleivg committed:
xrEngine: enable some checking on linux

eagleivg committed:
Common: fix some comment and printed messages

eagleivg committed:
xrEngine: implement splash on linux

eagleivg committed:
xrRender: fix load detail textures

eagleivg committed:
xrCore: fix delimiter in ini file loader

Zegeri committed:
xrAICore: Fix hash_fixed_vertex_manager::to_u32 by forward declaring

eagleivg committed:
xrEngine: fix dirty hack for processing of priority

eagleivg committed:
xrRenderGL: fix crash at exit from credits video

eagleivg committed:
xrGame: fix sound loading on Linux

eagleivg committed:
xrRenderPC_GL: fix texture loading

eagleivg committed:
xrGame: remname forgotten virtual methods in child classes

eagleivg committed:
xrCore: add platform name to log output

q4a committed:
Revert "xrAICore: Fix hash_fixed_vertex_manager::to_u32 by forward declaring"

This reverts commit 87807456d6e6ce229944a3769461816abae8e0bf.

eagleivg committed:
xrCore: fix load archived config files

Xottab-DUTY commented on src/xrCore/xrMemory.cpp:
```diff
@@ -1,5 +1,7 @@
 #include "stdafx.h"
 
+#include "SDL.h"
```
It's included in Common.hpp, please remove it from here.

eagleivg committed:
xrGame: fix undefined symbol: _ZN12SPHBonesDataC1Ev

eagleivg committed:
xrGame: fix some undefined reference

eagleivg committed:
xrGame: fix ALife resource loading

FreeZoneMods commented on src/Layers/xrRender/SkeletonAnimated.cpp:
```diff
@@ -771,6 +771,9 @@ void CKinematicsAnimated::Load(const char* N, IReader* data, u32 dwFlags)
         {
             data->r_stringZ(nm, sizeof(nm));
+#ifdef LINUX
+            while (char* sep = strchr(nm, '\\')) *sep = '/';
```
Looks like all this magic is necessary only when we open the file. So I suggest you to remove this IFDEF's from all code and encapsulate path corrections only in the classes which are responsible for working with the file system.

eagleivg committed:
COmmon, xrCore: refactor SDL2 includes

eagleivg committed:
Revert "COmmon, xrCore: refactor SDL2 includes"

This reverts commit 56931f3ccf9479730efab96226fcf7b42bf23e24.

Zegeri committed:
xrGame: Add ik/*.cxx files to cmake

Zegeri committed:
Fix CUIMainIngameWnd xml section loading

From the standard: "If a call to sprintf or snprintf causes copying to
take place between objects that overlap, the behavior is undefined."

Zegeri committed:
xrAICore: Fix hash_fixed_vertex_manager::to_u32 in Linux

Hopefully this time it doesn't break Windows.

Zegeri committed:
Use upper case in directory Localization in fsgame.ltx

Lazy solution for Linux's case-sensitive behaviour. A proper solution
might be required in the future.

Zegeri committed:
xrAICore: Remove -DAI_COMPILER in cmake.

This flag is supposed to be used when compiling the xrAI tool

Zegeri committed:
Don't use DELIMITER for textures defined in xml files with "\"

Zegeri committed:
Use mutex lock to avoid race condition when loading textures

eagleivg committed:
xrGame: fix some IK undefined reference

eagleivg committed:
Common: big refactor FS loading on linux

Zegeri committed:
xrUICore: Export color_animation symbol

Zegeri committed:
Use xrCore's locks instead of std::mutex

I suggest that in the future Lock should become a wrapper of std::mutex
and the current Lock should be renamed to RecursiveLock.

q4a committed:
Merge pull request #18 from Zegeri/clean

Fix windows build and add some minor fixes

FreeZoneMods commented on src/xrUICore/XML/UITextureMaster.cpp:
```diff
@@ -54,6 +56,8 @@ void CUITextureMaster::ParseShTexInfo(pcstr path, pcstr xml_file)
             /* avo: fix issue when values were not updated (silently skipped) when same key is encountered more than once. This is how std::map is designed. 
             /* Also used more efficient C++11 std::map::emplace method instead of outdated std::pair::make_pair */
             /* XXX: avo: note that xxx.insert(mk_pair(v1,v2)) pattern is used extensively throughout solution so there is a good potential for other bug fixes/improvements */
+            static Lock new_texture_lock;
```
Please check if the 'static' lock is enough. Maybe we need some sync mechanism for m_textures in the other places to avoid UB? Also maybe it's better to make the lock a class member?

eagleivg replied:
Im think that this place dont needed any sync, because in Windows no problem with textures. Most likely, the problem is elsewhere, because the game on linux immediately drops, and in different places.

FreeZoneMods commented on src/xrGame/xrServer.cpp:
```diff
@@ -986,7 +988,9 @@ void xrServer::create_direct_client()
     xr_strcpy(cl_data.name, "single_player");
-#ifndef LINUX // FIXME!!!
+#ifdef LINUX
+    cl_data.process_id = getpid();
```
Please consider moving such IFDEFs to separate function with platform-specific implementation. Let's won't clutter the code with IFDEFs.

eagleivg replied:
Sounds reasonable. Although the number of such platform-dependent calls is not very small, I think it’s time to reap the guidance of beauty after running the game on Linux

FreeZoneMods commented on src/xrGame/xrGame.cpp:
```diff
@@ -78,11 +78,14 @@ static void load(int argc, char** argv, char** envp)
 // XXX nitrocaster PROFILER: temporarily disabled due to linkage issues
 // g_profiler			= new CProfiler();
 #endif
+    gStringTable = new CStringTable();
```
Why is the whole functions 'load' and 'unload' are IFDEFed here? The code of the functions is identical with code in DllMain. So maybe it's better to call 'load' and 'unload' from DllMain to avoid code duplication?

eagleivg replied:
__attribute__((constructor)) and __attribute__((destructor)) will not be accepted by msvc, and will still be under #ifdef. To be honest, I just don’t know how to do it beautifully.

FreeZoneMods replied:
I am not sure about the behaviour of the compiler on Linux, but what about IFDEF'ing only the lines with __attribute (and leaving the function body not IFDEF'ed)? It would allow us to call the specified functions from DllMain on Windows instead of duplicating the code. At least such solution relieves us from maintaining sync for some parts of the code in case of future changes.

Zegeri replied:
You can replace `__attribute__((constructor))` with `[[gnu::constructor]]` to make MSVC accept those functions. Maybe it prints an "unknown attribute" warning, but it won't be an error.

FreeZoneMods commented on src/xrGame/ui/UIPdaWnd.cpp:
```diff
@@ -165,6 +165,7 @@ void CUIPdaWnd::Update()
     m_clock->TextItemControl().SetText(
         InventoryUtilities::GetGameTimeAsString(InventoryUtilities::etpTimeToMinutes).c_str());
 
+    assert(pUILogsWnd);
```
R_ASSERT?

FreeZoneMods commented on src/xrGame/ui/UIMainIngameWnd.cpp:
```diff
@@ -235,11 +235,10 @@ void CUIMainIngameWnd::Init()
-        CUIXmlInit::InitStatic(uiXml, path, 0, m_quick_slots_icons.back());
-        xr_sprintf(path, "%s:counter", path);
-        UIHelper::CreateStatic(uiXml, path, m_quick_slots_icons.back());
+        std::string path = "quick_slot" + std::to_string(i);
```
shared_str?

FreeZoneMods commented on src/xrGame/script_sound.cpp:
```diff
@@ -29,7 +29,7 @@ CScriptSound::CScriptSound(LPCSTR caSoundName, ESoundTypes sound_type)
 CScriptSound::~CScriptSound()
 {
     if (m_sound._feedback())
-        GEnv.ScriptEngine->script_log(LuaMessageType::Error, "Playing sound is not completed, but is destroying \"%s\"!", 
```
Dummy change :( No sense to change an empty space in independent file without any other changes. Such commit just makes the history more complicated (i.e for using blame feature)

FreeZoneMods commented on src/xrGame/ik/limb.cpp:
```diff
@@ -30,9 +30,13 @@
 #include "limb.h"
 
+#ifdef LINUX
+#undef roundup // defined in <sys/param.h>
```
Do you really need this ifdef? Looks suspicious.

FreeZoneMods commented on src/xrGame/console_commands.cpp:
```diff
@@ -156,11 +156,11 @@ static void full_memory_stats()
-    Msg("* [x-ray]: process heap[%u K]", _process_heap / 1024);
-    Msg("* [x-ray]: economy: strings[%d K], smem[%d K]", _eco_strings / 1024, _eco_smem);
+    Msg("* [ Render ]: textures[%d K]", (m_base + m_lmaps) / 1024);
+    Msg("* [ x-ray  ]: process heap[%u K]", _process_heap / 1024);
```
Strange style for logging. Do we really need the extra spaces with extra changes here?

FreeZoneMods commented on src/xrGame/atlas_submit_queue.cpp:
```diff
@@ -120,8 +120,9 @@ void atlas_submit_queue::do_atlas_reward(
 
     m_atlas_in_process = true;
-
+#ifdef WINDOWS
```
Maybe mark it as TODO?

FreeZoneMods commented on src/xrGame/Level.cpp:
```diff
@@ -551,6 +554,7 @@ void CLevel::OnFrame()
     {
         if (g_mt_config.test(mtLevelSounds))
         {
+            assert(m_level_sound_manager);
```
R_ASSERT?

FreeZoneMods commented on src/xrGame/Level.cpp:
```diff
@@ -442,7 +442,10 @@ void CLevel::OnFrame()
     {
         if (g_mt_config.test(mtMap))
+        {
+            assert(m_map_manager);
```
R_ASSERT?

FreeZoneMods commented on src/xrGame/GameObject.cpp:
```diff
@@ -112,10 +112,13 @@ void CGameObject::MakeMeCrow()
+#else
+     if (__sync_val_compare_and_swap((long*)&dwFrame_AsCrow, object_frame_id, device_frame_id) == device_frame_id)
 #endif
+        return;
```
Return here looks odd and makes the execution logic less clear. Consider rewriting it.

FreeZoneMods commented on src/xrGame/GameObject.cpp:
```diff
@@ -112,10 +112,13 @@ void CGameObject::MakeMeCrow()
     u32 const device_frame_id = Device.dwFrame;
     u32 const object_frame_id = dwFrame_AsCrow;
-#ifndef LINUX // FIXME!!!
+#ifndef LINUX
```
Consider moving the code of IFDEF to separate function with platform-dependent implementation.

FreeZoneMods commented on src/xrGame/CharacterPhysicsSupport.h:
```diff
@@ -153,9 +153,7 @@ class CCharacterPhysicsSupport : public CPHSkeleton, public CPHDestroyable
-    {
-        R_ASSERT2(false, "Can not assign it");
-    }
+        = delete;
```
Why this placed on the separate line? I think it reduces readability of the code.

FreeZoneMods commented on src/xrEngine/splash.cpp:
```diff
@@ -2,24 +2,29 @@
 #include "xr_3da/resource.h"
 #include "splash.h"
 
+#if defined(WINDOWS)
```
Please consider reworking without IFDEFs.

FreeZoneMods commented on src/xrCore/xr_ini.cpp:
```diff
@@ -642,6 +635,9 @@ bool CInifile::save_as(pcstr new_fname)
 
     R_ASSERT(m_file_name && m_file_name[0]);
+#ifdef LINUX
+    while (char* sep = strchr(m_file_name, '\\')) *sep = '/';
```
Please consider moving this code to separate function with platform-specific implementation instead of IFDEFing it here directly.

FreeZoneMods commented on src/xrCore/file_stream_reader.cpp:
```diff
@@ -17,7 +17,10 @@ void CFileStreamReader::construct(LPCSTR file_name, const u32& window_size)
-    m_file_handle = ::open(file_name, O_RDONLY);
+    char path[PATH_MAX] = { 0 };
+    strcpy(path, file_name);
+    while (char* sep = strchr(path, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation instead of IFDEFing it here directly. Also it allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/fastdelegate.h:
```diff
@@ -52,6 +52,7 @@
 #endif // _MSC_VER > 1000
 
 #include <memory.h> // to allow <,> comparisons
+#include <cassert>
```
Why do you use it instead of R_ASSERT?

FreeZoneMods commented on src/xrCore/LocatorAPI.cpp:
```diff
@@ -661,6 +661,9 @@ bool CLocatorAPI::Recurse(pcstr path)
     glob_t globbuf;
 
+    while (char* sep = strchr(scanPath, '\\'))
+        *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation instead of IFDEFing it here directly. Also it allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/LocatorAPI.cpp:
```diff
@@ -632,6 +631,7 @@ bool ignore_path(const char* _path)
     else
         return true;
 #elif defined(LINUX)
+    while (char* sep = strchr((char *)_path, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/LocatorAPI.cpp:
```diff
@@ -462,6 +459,7 @@ void CLocatorAPI::archive::open()
     if (hSrcFile)
         return;
 
+    while (char* sep = strchr((char *)*path, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/FS.cpp:
```diff
@@ -111,6 +114,7 @@ static int open_internal(LPCSTR fn, int& handle)
 #if defined(WINDOWS)
     return (_sopen_s(&handle, fn, _O_RDONLY | _O_BINARY, _SH_DENYNO, _S_IREAD));
 #elif defined(LINUX)
+    while (char* sep = strchr((char *)fn, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/FS.cpp:
```diff
@@ -520,6 +524,7 @@ CVirtualFileRW::CVirtualFileRW(const char* cFileName)
     data = (char*)MapViewOfFile(hSrcMap, FILE_MAP_ALL_ACCESS, 0, 0, 0);
     R_ASSERT3(data, cFileName, xrDebug::ErrorToString(GetLastError()));
 #elif defined(LINUX)
+    while (char* sep = strchr((char *)cFileName, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/FS.cpp:
```diff
@@ -563,6 +568,7 @@ CVirtualFileReader::CVirtualFileReader(const char* cFileName)
 
     data = (char*)MapViewOfFile(hSrcMap, FILE_MAP_READ, 0, 0, 0);
 #elif defined(LINUX)
+    while (char* sep = strchr((char *)cFileName, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/xrCore/FS.cpp:
```diff
@@ -88,6 +88,9 @@ void VerifyPath(LPCSTR path)
         _mkdir(tmp);
     }
+#ifdef LINUX
+    while (char* sep = strchr((char *)path, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/Layers/xrRenderPC_GL/rgl.cpp:
```diff
@@ -452,6 +452,7 @@ void CRender::OnFrame()
     if (ps_r2_ls_flags.test(R2FLAG_EXP_MT_CALC))
     {
         // MT-details (@front)
+        assert (Details);
```
R_ASSERT?

FreeZoneMods commented on src/Layers/xrRenderPC_GL/glResourceManager_Resources.cpp:
```diff
@@ -512,6 +512,7 @@ CTexture* CResourceManager::_CreateTexture(LPCSTR _Name)
     R_ASSERT (_Name && _Name[0]);
     string_path Name;
     strcpy_s(Name, _Name); //. andy if (strext(Name)) *strext(Name)=0;
+
```
Unneccessary changes

FreeZoneMods commented on src/Common/PlatformLinux.inl:
```diff
@@ -333,7 +351,12 @@ inline int _filelength(int fd)
-#define _mkdir(dir) mkdir(dir, S_IRWXU)
+inline int _mkdir(const char *dir)
+{
+    while (char* sep = strchr((char *)dir, '\\')) *sep = '/';
```
Please consider moving this code (slashes conversion) to separate function (named like ConvertGamePathToPlatformSpecificPath) with platform-specific implementation. It makes the code more clear and allows to reduce code duplication.

FreeZoneMods commented on src/Common/FSMacros.hpp:
```diff
@@ -1,12 +1,12 @@
-#define _DELIMITER '/' //for looking
-#define DELIMITER "/" // for insert
-#elif defined(WINDOWS)
+//#if defined(LINUX)
```
Remove dead indefs?

FreeZoneMods commented on Externals/cximage/ximaenc.cpp:
```diff
@@ -996,7 +996,7 @@ bool CxImage::CheckFormat(CxFile * hFile, DWORD imagetype)
 bool CxImage::CheckFormat(BYTE * buffer, DWORD size, DWORD imagetype)
 {
-	if (buffer==NULL || size==NULL){
+	if (buffer==NULL || size==0){
```
nullptr
